library/core/src/slice/specialize.rs

```diff
@@ -0,0 +1,58 @@
+use crate::ptr::write_bytes;
+
+pub(super) trait SpecFill<T> {
+    fn spec_fill(&mut self, value: T);
+}
+
+impl<T: Clone> SpecFill<T> for [T] {
+    default fn spec_fill(&mut self, value: T) {
+        if let Some((last, elems)) = self.split_last_mut() {
+            for el in elems {
+                el.clone_from(&value);
+            }
+
+            *last = value
+        }
+    }
+}
+
+impl<T: Copy> SpecFill<T> for [T] {
+    default fn spec_fill(&mut self, value: T) {
+        for item in self.iter_mut() {
+            *item = value;
+        }
+    }
+}
+
+impl SpecFill<u8> for [u8] {
```

Comment:
The only difference when not specialize for `[u8]` is that compiler emits a check whether the slice is empty.
<https://rust.godbolt.org/z/z9fbcT>.

Maybe it is not worth to keep those specs for u8/i8/bool slice.